src/v1/entity/utils/validateKeyInput.ts

```diff
@@ -1,27 +1,138 @@
+}
+
+export class UnrecognizedKeyInputPropertyError extends Error {
+  constructor({ path }: { path?: string }) {
+    super(
+      `Unrecognized key input property${getPathMessage(path)}. Property is not tagged as key input.`
+    )
+  }
+}
+
+export class MissingRequiredPropertyError extends Error {
+  constructor({ path }: { path?: string }) {
+    super(`Missing always required key input property${getPathMessage(path)}.`)
+  }
+}
+
+export class UnexpectedPropertyError extends Error {
+  constructor({ path }: { path?: string }) {
+    super(`Unexpected key input property${getPathMessage(path)}.`)
+  }
+}
+
+type KeyInputValidator = <E extends EntityV2 | Item | Property>(
   entity: E,
-  keyInput: Record<string, any>
-) => keyInput is K
+  keyInput: KeyInput<E>,
+  path?: string
+) => void
+
+const validateProperties = (
+  itemOrMapped: Item | Mapped,
+  keyInput: ResolvedProperty,
+  path?: string
+): void => {
+  if (!isObject(keyInput))
+    throw new InvalidKeyInputValueTypeError({ expectedType: 'map', keyInput, path })
+
+  // Check that keyInput values match item or mapped
+  Object.entries(keyInput).forEach(([propertyName, propertyInput]) => {
+    const property = itemOrMapped._properties[propertyName]
+    // TODO, create joinPath util
+    const propertyPath = [path, propertyName].filter(Boolean).join('.')
+
+    if (property !== undefined) {
+      validateKeyInput(property, propertyInput, propertyPath)
+    } else {
+      if (isClosed(itemOrMapped)) {
+        throw new UnexpectedPropertyError({ path: propertyPath })
+      }
+      // TODO: create validateAny ?
+    }
+  })
+
+  // Check that all key & always required properties are present in keyInput
+  Object.entries(itemOrMapped._properties)
+    .filter(([, property]) => isKey(property) && property._required === 'always')
+    .forEach(([propertyName, property]) => {
+      const propertyKeyInput = keyInput[propertyName]
+      // TODO, create joinPath util
+      const propertyPath = [path, propertyName].filter(Boolean).join('.')
+
+      if (propertyKeyInput !== undefined) {
+        validateKeyInput(property, propertyKeyInput, propertyPath)
+      } else {
+        throw new MissingRequiredPropertyError({ path: propertyPath })
+      }
+    })
+}
 
 /**
  * Validates the primary key input of a single item command (GET, DELETE ...) for a given Entity
  *
- * @param entity Entity
+ * @param entry Entity | Item | Property
  * @param keyInput Key input
- * @return Boolean
+ * @param path _(optional)_ Path of the property in the related item (string)
+ * @return void
  */
-export const validateKeyInput: KeyInputValidator = <
-  E extends EntityV2,
-  K extends Record<string, any> = KeyInput<E>
->(
-  entity: E,
-  keyInput: Record<string, any>
-): keyInput is K => {
-  entity
-  keyInput
-  // TODO
-  return true
+export const validateKeyInput: KeyInputValidator = <E extends EntityV2 | Item | Property>(
+  entry: E,
+  keyInput: KeyInput<E>,
+  path?: string
+): void => {
```

Comment:
imo we shouldn't use a single character for generics,
wdyt about renaming E to Entry?

Reply:
Yes I was planning to do that on all codebase in next PR